### PR TITLE
Permissioning - A better solution to fix NullPointerException on createPayload

### PR DESCRIPTION
**Pull request by camilodotto**
## PR description
This is a more elegant solution then PR 5374 for issue 5370. When running on docker, URI returns null on getHost() method but returns the correct hostname on toString() method, owning to the 'underline character' on hostname in docker-compose services.

## Fixed Issue(s)
#5370

### CODE REVIEW

**Comment by non-fungible-nelson** [comment]
Thanks for this! Will look to review

File: ethereum/permissioning/src/main/java/org/hyperledger/besu/ethereum/permissioning/NodeSmartContractV2PermissioningController.java [hunk 1]
```diff
@@ -113,6 +114,23 @@ private Bytes createPayload(final EnodeURL enodeUrl) {
     }
   }
 
+  private String getHost(final EnodeURL enodeUrl) {
```
Comment (macfarla) [inline ethereum/permissioning/src/main/java/org/hyperledger/besu/ethereum/permissioning/NodeSmartContractV2PermissioningController.java:117]: would love to see some unit tests for this method - and it should live with the EnodeURL class (EnodeURLImpl) - can see this being useful elsewhere

Comment (camilodotto) [reply ethereum/permissioning/src/main/java/org/hyperledger/besu/ethereum/permissioning/NodeSmartContractV2PermissioningController.java]: @macfarla we did some unit tests and moved getHost() to EnodeURL

Comment (camilodotto) [reply ethereum/permissioning/src/main/java/org/hyperledger/besu/ethereum/permissioning/NodeSmartContractV2PermissioningController.java]: https://stackoverflow.com/questions/25993225/uri-gethost-returns-null-why

Comment (non-fungible-nelson) [reply ethereum/permissioning/src/main/java/org/hyperledger/besu/ethereum/permissioning/NodeSmartContractV2PermissioningController.java]: I differ to @macfarla on this. She can review your PR. Have you reviewed the failing tests? You can possibly ignore the plugins Checksum issue.

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/peers/EnodeURLImplTest.java [hunk 2]
```diff
@@ -967,4 +969,99 @@ public void toURIWithoutDiscoveryPortShouldKeepDomainWhenFound() {
...
+            .nodeId(VALID_NODE_ID)
+            .ipAddress(IPV6_FULL_ADDRESS)
+            .disableListening()
+            .discoveryPort(DISCOVERY_PORT)
+            .build();
+    assertEquals(IPV6_COMPACT_ADDRESS, enode.getHost());
+  }
+
+  @Test
+  public void getHostShoulReturnHostNameWhenHostContainsUnderscore() {
```
Comment (macfarla) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/peers/EnodeURLImplTest.java:1010]:
```suggestion
  public void getHostShouldReturnHostNameWhenHostContainsUnderscore() {
```

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/peers/EnodeURLImplTest.java [hunk 3]
```diff
@@ -967,4 +969,99 @@ public void toURIWithoutDiscoveryPortShouldKeepDomainWhenFound() {
...
+    Mockito.when(enodeMock.toURIWithoutDiscoveryPort()).thenReturn(uriMock);
+    Mockito.when(uriMock.getHost()).thenReturn(null);
+    String hostName = "host_name";
+    Mockito.when(uriMock.toString())
+        .thenReturn("enode://" + VALID_NODE_ID + "@" + hostName + ":9999");
+    assertEquals(hostName, enodeMock.getHost());
+  }
+
+  @Test
+  public void getHostShoulReturnEmptyStringWhenUriToStringDoesNotContainAt() {
```
Comment (macfarla) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/peers/EnodeURLImplTest.java:1023]:
```suggestion
  public void getHostShouldReturnEmptyStringWhenUriToStringDoesNotContainAt() {
```

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/peers/EnodeURLImplTest.java [hunk 4]
```diff
@@ -967,4 +969,99 @@ public void toURIWithoutDiscoveryPortShouldKeepDomainWhenFound() {
...
+    Mockito.when(enodeMock.getHost()).thenCallRealMethod();
+    Mockito.when(enodeMock.toURIWithoutDiscoveryPort()).thenReturn(uriMock);
+    Mockito.when(uriMock.getHost()).thenReturn(null);
+    String hostName = "host_name";
+    Mockito.when(uriMock.toString()).thenReturn("enode://" + VALID_NODE_ID + hostName + ":9999");
+    assertEquals("", enodeMock.getHost());
+  }
+
+  @Test
+  public void getHostShoulReturnEmptyStringWhenUriToStringReturnsColonBeforeAt() {
```
Comment (macfarla) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/peers/EnodeURLImplTest.java:1035]:
```suggestion
  public void getHostShouldReturnEmptyStringWhenUriToStringReturnsColonBeforeAt() {
```

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/peers/EnodeURLImplTest.java [hunk 5]
```diff
@@ -967,4 +969,99 @@ public void toURIWithoutDiscoveryPortShouldKeepDomainWhenFound() {
...
+    Mockito.when(enodeMock.getHost()).thenCallRealMethod();
+    Mockito.when(enodeMock.toURIWithoutDiscoveryPort()).thenReturn(uriMock);
+    Mockito.when(uriMock.getHost()).thenReturn(null);
+    String hostName = "host_name";
+    Mockito.when(uriMock.toString()).thenReturn("enode://" + VALID_NODE_ID + "@" + hostName);
+    assertEquals("", enodeMock.getHost());
+  }
+
+  @Test
+  public void getHostShoulReturnHostNameDomainFound() {
```
Comment (macfarla) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/peers/EnodeURLImplTest.java:1047]:
```suggestion
  public void getHostShouldReturnHostNameWhenDomainFound() {
```

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/peers/EnodeURLImplTest.java [hunk 6]
```diff
@@ -967,4 +969,99 @@ public void toURIWithoutDiscoveryPortShouldKeepDomainWhenFound() {
...
+    String enodeString = String.format("enode://%s@%s:%d", VALID_NODE_ID, site, 9999);
+    final EnodeURL enode =
+        EnodeURLImpl.fromString(
+            enodeString,
+            ImmutableEnodeDnsConfiguration.builder().dnsEnabled(true).updateEnabled(true).build());
+    assertEquals(site, enode.getHost());
+  }
+
+  @Test
+  public void getHostShoulReturnLoopBackWhenHostNameDomainNotFound() {
```
Comment (macfarla) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/peers/EnodeURLImplTest.java:1058]:
```suggestion
  public void getHostShouldReturnLoopBackWhenDomainNotFound() {
```

**Review by macfarla — APPROVED** [review]
LGTM

**Pull request merged**